Job resource - kubectl part

soltysh opened the pull request:
This is 2nd part of the work for the job controller. Depends on #12910.

@bgrant0607 @erictune @smarterclayton ptal
@derekwaynecarr @sdminonne @akaram fyi

soltysh committed:
Job clients, printer and describer

soltysh commented:
Appreciate info about what failed actually?

kargakis commented:
@kubernetes/kubectl

jlowdermilk commented:
jenkins failure looks like a flake. kicking

jlowdermilk commented:
@k8s-bot test this

soltysh commented:
@jlowdermilk apparently still something is wrong, could you send over what's the problem. The only resource I can preview is the test artifacts but these don't provide me information about test failure :( Appreciate that.

erictune commented on pkg/expapi/types.go:
```diff
@@ -319,3 +319,98 @@ type DaemonList struct {
+	Conditions []JobCondition `json:"conditions,omitempty"`
+
+	// CreationTime represents time when the job was created
+	CreationTime util.Time `json:"creationTime,omitempty"`
```
How would this differ from metadata.CreationTimestamp?

soltysh replied:
The general goal here is to track the actual creation of a job execution, it is intended for future tracking of how long the job was running. But given that we don't distinguish pending from running pods I guess that can be easily replaced with the StartTime below. I'll remove it then.

erictune commented on pkg/expapi/types.go:
```diff
@@ -319,3 +319,98 @@ type DaemonList struct {
+	CreationTime util.Time `json:"creationTime,omitempty"`
+
+	// StartTime represents time when the job was started
+	StartTime util.Time `json:"startTime,omitempty"`
```
should this be called StartTimestamp for consistency, and likewise below?
@bgrant0607

soltysh replied:
According to [naming conventions](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#naming-conventions): The name of a field that specifies the time at which something occurs should be called somethingTime. Do not use stamp.

bgrant0607 replied:
@soltysh is correct. startTime. We do this at least 3 different ways in the API today, so consistency in the current API version is a lost cause. I chose the convention I liked best.

erictune commented on pkg/expapi/types.go:
```diff
@@ -319,3 +319,98 @@ type DaemonList struct {
+	// CreationTime represents time when the job was created
+	CreationTime util.Time `json:"creationTime,omitempty"`
+
+	// StartTime represents time when the job was started
```
Document similar to `CreationTimestamp` in `type ObjectMeta`.
e.g.:
```
        // It is not guaranteed to be set in happens-before order across separate operations.
        // Clients may not set this value. It is represented in RFC3339 form and is in UTC.
        //
        // Populated by the system.
        // Read-only.
```

soltysh replied:
It also makes sense to do that for CompletionTime.

erictune commented on pkg/expapi/types.go:
```diff
@@ -319,3 +319,98 @@ type DaemonList struct {
+	// Active is the number of actively running pods.
+	Active int `json:"active,omitempty"`
+
+	// Successful is the number of pods successfully completed their job.
```
Sucessful is the number of pods which reached Phase Succeeded.

bgrant0607 replied:
That sounds right to me, for the current v1 Pod API.

soltysh replied:
Done.

erictune commented on pkg/expapi/types.go:
```diff
@@ -319,3 +319,98 @@ type DaemonList struct {
+	Unsuccessful int `json:"unsuccessful,omitempty"`
+}
+
+// JobStatusResult is a wrapper for JobStatus returned by kubelet that can be encode/decoded
```
Kubelet is going to return JobStatus?  Why does Kubelet know about that?  Or is this a copy/paste leftover, and the real reason for this is that it is for the /status resource?

bgrant0607 replied:
I hope that's a typo. Kubelet shouldn't know anything about jobs.

soltysh replied:
It was copy/paste leftover from some other place, thanks for spotting that!

erictune commented on pkg/expapi/validation/validation.go:
```diff
@@ -156,3 +158,50 @@ func ValidateDaemonSpec(spec *expapi.DaemonSpec) errs.ValidationErrorList {
+
+	selector := labels.Set(spec.Selector).AsSelector()
+	if selector.Empty() {
+		allErrs = append(allErrs, errs.NewFieldRequired("selector"))
```
We talked about autogenerating selector if not specified.  Did we decide not to do that?  Making it optional later would be a breaking API change, IIUC.

erictune replied:
I guess you are making it as similar to replication controller as possible.  So, I guess that is fine.

soltysh replied:
I thought the auto-generating selector was regarding the scheduled jobs at the time when both of the proposals were submitted as one PR. And yes, I'm trying to make jobs similar to RC, because they are meant to be supplementary to RC, imho.

soltysh commented:
@erictune @bgrant0607 since your comments where regarding the job resource I've addressed them in #12910.

erictune commented:
I'm on vacation this week (:house: :mouse:) and will take a look next week.

kargakis commented on pkg/expapi/v1/types.go:
```diff
@@ -318,3 +318,98 @@ type DaemonList struct {
+	Status JobStatus `json:"status,omitempty"`
+}
+
+type JobConditionType string
```
godoc

kargakis commented on pkg/registry/job/etcd/etcd.go:
```diff
@@ -0,0 +1,77 @@
+	"k8s.io/kubernetes/pkg/storage"
+)
+
+// rest implements a RESTStorage for jobs against etcd
```
// REST

kargakis commented:
First pass looks good

kargakis commented:
Leaving this comment here after discussing with @soltysh on IRC, we will need to implement the Reaper for graceful deleting a Job and maybe the Scaler?

soltysh commented:
Rebased against latest master to pick up the client changes, esp. the experimental part.

smarterclayton commented:
Yes, need reaper and scaler.

kargakis commented:
> Yes, need reaper and scaler.

Maybe scaling can be handled by the job controller like [it will happen](https://github.com/mikedanese/kubernetes/commit/7fe94f46797d2aec69262b3e0e2accf78a94e05d#diff-2076b6236060a0cca0937ac2907ab263R214) with daemons ?

soltysh commented:
> Yes, need reaper and scaler.

It does make sense to have both, it'll be easier to modify parallelism with scale, that's how I imagine it to work.

erictune commented on pkg/client/unversioned/jobs.go:
```diff
@@ -0,0 +1,96 @@
+// List returns a list of jobs that match the label and field selectors.
+func (c *jobs) List(label labels.Selector) (result *expapi.JobList, err error) {
+	result = &expapi.JobList{}
+	err = c.r.Get().Namespace(c.ns).Resource("jobs").LabelsSelectorParam(label).Do().Into(result)
```
Follow the pattern of `func (*pods) List(...)` in `pkg/client/unversioned/pods.go`.  That takes both a label selector and a field selector argument.  The fact that there is not a field selector for replicationControllers is a bug.

erictune commented on pkg/client/unversioned/jobs.go:
```diff
@@ -0,0 +1,96 @@
+}
+
+// Delete deletes a job, returns error if one occurs.
+func (c *jobs) Delete(name string) (err error) {
```
This should follow the pattern of pod and accept DeleteOptions.

erictune replied:
I think it is okay if it initially does not respect them, but better to get the prototype right now.

erictune commented on pkg/client/unversioned/jobs.go:
```diff
@@ -0,0 +1,96 @@
+		LabelsSelectorParam(label).
+		FieldsSelectorParam(field).
+		Watch()
+}
```
Do you want to just add `UpdateStatus` here?  It is not much extra code.  Just copy what is done in pkg/client/unversioned/pods.go

erictune commented:
looks good aside from comments above.

soltysh commented:
@erictune I've addressed your comments. The PR will fail current checks, I'll wait with rebasing to get the API in, first. The only question that is left is, should I implement the reaper and scaler now or can it be left afterwards?

soltysh commented:
Today during demoing job controller, I've realized I'm not showing Completions and Parallelism, added those to describer.

erictune commented:
@karlkfi  It seems we do not do automatic generation of static docs for experimental api group.   We should.  Filed https://github.com/kubernetes/kubernetes/issues/13883 for that.

mikedanese commented:
@wojtek-t would you mind reviewing the apiserver/registry portion of this PR?

smarterclayton commented on pkg/registry/job/rest.go:
```diff
@@ -0,0 +1,106 @@
+// PrepareForUpdate clears fields that are not allowed to be set by end users on update.
+func (jobStrategy) PrepareForUpdate(obj, old runtime.Object) {
+	// TODO: once jobs have a sub status sub-resource we can enable this.
+	//newController := obj.(*expapi.Job)
```
This needs to be fixed now that you have statis

soltysh replied:
This is from the new resouce commit. It'll be fixed once I rebase this PR on top of current master.

soltysh commented:
I've rebased this PR to pick up changes from job resource. PTAL.

erictune commented:
LGTM.  Please squash.

erictune commented:
Hoping jenkins was a flake and that squashing will trigger a rerun which will pass.

soltysh commented:
Yeah, I'm hoping that as well :wink: All commits squashed.

mikedanese commented:
This needs a rebase (and you will have to fix compilation issue after rebasing)

soltysh commented:
Rebased, fixed the compilation issue @mikedanese mentioned, should be good.

smarterclayton commented:
Squash and rebase looks clean to me, reapplying LGTM

smarterclayton commented:
ok to test
